pkg/k8sutil/k8sutil.go

```diff
@@ -175,20 +176,33 @@ func UpdateStatefulSet(ctx context.Context, sstClient clientappsv1.StatefulSetIn
 	return nil
 }
 
-// UpdateSecret merges metadata of existing Secret with new one and updates it.
-func UpdateSecret(ctx context.Context, secretClient clientv1.SecretInterface, secret *v1.Secret) error {
-	existingSecret, err := secretClient.Get(ctx, secret.Name, metav1.GetOptions{})
-	if err != nil {
-		return errors.Wrap(err, "getting secret object failed")
-	}
-
-	mergeMetadata(&secret.ObjectMeta, existingSecret.ObjectMeta)
-
-	_, err = secretClient.Update(ctx, secret, metav1.UpdateOptions{})
+// CreateOrUpdateSecret merges metadata of existing Secret with new one and updates it.
+func CreateOrUpdateSecret(ctx context.Context, secretClient clientv1.SecretInterface, desired *v1.Secret) error {
+	existingSecret, err := secretClient.Get(ctx, desired.Name, metav1.GetOptions{})
 	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(
+				err,
+				"failed to check whether tls assets secret already exists for Prometheus %v in namespace %v",
+				desired.Name,
+				desired.Namespace,
+			)
+		}
+		_, err = secretClient.Create(ctx, desired, metav1.CreateOptions{})
 		return err
 	}
-
+	// If secret exist,
+	mutated := existingSecret.DeepCopyObject().(*v1.Secret)
+
+	mergeMetadata(&desired.ObjectMeta, mutated.ObjectMeta)
+	mutated.Data = desired.Data
+	// diff and whether update
+	if !apiequality.Semantic.DeepEqual(existingSecret, mutated) {
```

Comment:
why not reflect.DeepEqual() on the Data fields?

Reply:
We also need to compare labels and annoations. `apiequality.Semantic.DeepEqual` method is like reflect.DeepEqual, but focused on semantic equality instead of memory equality.

Reply:
:+1: looking quickly at the definition of `apiequality.Semantic.DeepEqual()`, I thought that it was only checking a few fields but now I see that it uses reflect.DeepEqual() when needed underneath :)